File: go-sdk/kestra_api_client/notifications_api.go
package kestra_api_client

import (
	"context"
	"time"
)

// NotificationsAPI covers /api/v1/notifications — the in-app notifications of
// the currently authenticated user. These paths carry no tenant segment; the
// server resolves the current user and their accessible tenants.
type NotificationsAPI struct {
	baseAPI
}

// ApiNotification is a single in-app notification.
type ApiNotification struct {
	Id          string     `json:"id"`
	UserId      string     `json:"userId,omitempty"`
	TenantId    string     `json:"tenantId,omitempty"`
	Type        string     `json:"type,omitempty"`
	Title       string     `json:"title,omitempty"`
	ReferenceId string     `json:"referenceId,omitempty"`
	Current     *int       `json:"current,omitempty"`
	Total       *int       `json:"total,omitempty"`
	Read        bool       `json:"read"`
	CreatedDate *time.Time `json:"createdDate,omitempty"`
	UpdatedDate *time.Time `json:"updatedDate,omitempty"`
}

func (o *ApiNotification) GetId() string    { return o.Id }
func (o *ApiNotification) GetRead() bool    { return o.Read }
func (o *ApiNotification) GetTitle() string { return o.Title }

// ApiNotificationsSince is the response of GET /api/v1/notifications/since.
type ApiNotificationsSince struct {
	Notifications []ApiNotification `json:"notifications,omitempty"`
	ServerTime    *time.Time        `json:"serverTime,omitempty"`
}

func (o *ApiNotificationsSince) GetNotifications() []ApiNotification { return o.Notifications }

// ApiNotificationHistory is the response of GET /api/v1/notifications/history.
// It is cursor-paginated: pass NextCursor back as `before` to page further.
type ApiNotificationHistory struct {
	Notifications []ApiNotification `json:"notifications,omitempty"`
	ServerTime    *time.Time        `json:"serverTime,omitempty"`
	NextCursor    string            `json:"nextCursor,omitempty"`
}

func (o *ApiNotificationHistory) GetNotifications() []ApiNotification { return o.Notifications }
func (o *ApiNotificationHistory) GetNextCursor() string               { return o.NextCursor }

// ApiMarkAllRead is the response of POST /api/v1/notifications/read-all.
type ApiMarkAllRead struct {
	Updated int `json:"updated"`
}

func (o *ApiMarkAllRead) GetUpdated() int { return o.Updated }

// NotificationsSince returns notifications created since the given instant.
// Backs GET /api/v1/notifications/since.
func (a *NotificationsAPI) NotificationsSince(ctx context.Context, since time.Time) (*ApiNotificationsSince, error) {
	params := buildQueryParams("since", since.UTC().Format(time.RFC3339))
	return doJSON[*ApiNotificationsSince](&a.baseAPI, ctx, "GET", superadminPath("notifications", "since"), nil, params)
}

// NotificationHistory returns a cursor-paginated page of the current user's
// notification history. `before` is an opaque cursor (nil for the first page);
// `limit` defaults to 20 server-side when nil. Backs GET /api/v1/notifications/history.
func (a *NotificationsAPI) NotificationHistory(ctx context.Context, before *string, limit *int) (*ApiNotificationHistory, error) {
	params := buildQueryParams("before", before, "limit", limit)
	return doJSON[*ApiNotificationHistory](&a.baseAPI, ctx, "GET", superadminPath("notifications", "history"), nil, params)
}

// UnreadNotificationCount returns the number of unread notifications for the
// current user. Backs GET /api/v1/notifications/unread-count.
func (a *NotificationsAPI) UnreadNotificationCount(ctx context.Context) (int64, error) {
	return doJSON[int64](&a.baseAPI, ctx, "GET", superadminPath("notifications", "unread-count"), nil, nil)
}

// MarkNotificationRead marks a notification read. Idempotent. Backs POST
// /api/v1/notifications/{id}/read.
func (a *NotificationsAPI) MarkNotificationRead(ctx context.Context, id string) error {
	return a.doVoidJSON(ctx, "POST", superadminPath("notifications", id, "read"), nil, nil)
}

// MarkNotificationUnread marks a notification unread. Idempotent. Backs POST
// /api/v1/notifications/{id}/unread.
func (a *NotificationsAPI) MarkNotificationUnread(ctx context.Context, id string) error {
	return a.doVoidJSON(ctx, "POST", superadminPath("notifications", id, "unread"), nil, nil)
}

// MarkAllNotificationsRead marks every notification of the current user read.
// Backs POST /api/v1/notifications/read-all.
func (a *NotificationsAPI) MarkAllNotificationsRead(ctx context.Context) (*ApiMarkAllRead, error) {
	return doJSON[*ApiMarkAllRead](&a.baseAPI, ctx, "POST", superadminPath("notifications", "read-all"), nil, nil)
}

File: python/python-sdk/kestrapy/api/notifications_api.py
from datetime import datetime, timezone
from typing import Optional

from kestrapy.base_api import BaseApi
from kestrapy.models.notification_controller_api_mark_all_read import (
    NotificationControllerApiMarkAllRead,
)
from kestrapy.models.notification_controller_api_notification_history import (
    NotificationControllerApiNotificationHistory,
)
from kestrapy.models.notification_controller_api_notifications_since import (
    NotificationControllerApiNotificationsSince,
)


class NotificationsApi(BaseApi):
    """In-app notifications of the current user (`/api/v1/notifications`).

    These paths carry no tenant segment; the server resolves the current user
    and their accessible tenants.
    """

    def notifications_since(self, since: datetime) -> NotificationControllerApiNotificationsSince:
        """Return notifications created since the given instant.
        Backs GET /api/v1/notifications/since."""
        if since.tzinfo is None:
            since = since.replace(tzinfo=timezone.utc)
        stamp = since.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
        params = self._build_query_params(since=stamp)
        path = self._superadmin_path("notifications", "since")
        return self._json_request("GET", path, NotificationControllerApiNotificationsSince, params=params)

    def notification_history(
        self, before: Optional[str] = None, limit: Optional[int] = None
    ) -> NotificationControllerApiNotificationHistory:
        """Return a cursor-paginated page of the current user's notification
        history. ``before`` is an opaque cursor (None for the first page);
        ``limit`` defaults to 20 server-side. Backs GET /api/v1/notifications/history."""
        params = self._build_query_params(before=before, limit=limit)
        path = self._superadmin_path("notifications", "history")
        return self._json_request("GET", path, NotificationControllerApiNotificationHistory, params=params)

    def unread_notification_count(self) -> int:
        """Return the number of unread notifications for the current user.
        Backs GET /api/v1/notifications/unread-count."""
        path = self._superadmin_path("notifications", "unread-count")
        return self._json_request("GET", path, int)

    def mark_notification_read(self, id: str) -> None:
        """Mark a notification read (idempotent). Backs POST /api/v1/notifications/{id}/read."""
        path = self._superadmin_path("notifications", id, "read")
        self._void_request("POST", path)

    def mark_notification_unread(self, id: str) -> None:
        """Mark a notification unread (idempotent). Backs POST /api/v1/notifications/{id}/unread."""
        path = self._superadmin_path("notifications", id, "unread")
        self._void_request("POST", path)

    def mark_all_notifications_read(self) -> NotificationControllerApiMarkAllRead:
        """Mark every notification of the current user read.
        Backs POST /api/v1/notifications/read-all."""
        path = self._superadmin_path("notifications", "read-all")
        return self._json_request("POST", path, NotificationControllerApiMarkAllRead)
